universal-application-tool-0.0.1/app/services/applicant/predicate/PredicateExpressionNode.java

```diff
@@ -0,0 +1,28 @@
+package services.applicant.predicate;
+
+import javax.annotation.Nullable;
+
+/** Represents a predicate that can be evaluated over {@link services.applicant.ApplicantData}. */
+public class PredicateExpressionNode {
+
+  private final PredicateExpressionNodeType type;
+  private final LeafOperationExpressionNode leafNode;
+
+  public PredicateExpressionNode(
+      PredicateExpressionNodeType type, @Nullable LeafOperationExpressionNode leafNode) {
+    this.type = type;
+    this.leafNode = leafNode;
```

Comment:
suggestion (blocking): It'd be best if the node param were generic and not nullable. An alternative approach could be to have a `interface ConcretePredicateExpressionNode` that `Leaf...`, `And...`, and `Or...` nodes all implement with a single required method `getType()` that returns the type enum for the node. This wrapper class would initialize with just that one argument and perform checking and casting in each of its `get*Node()` methods.

Reply:
Thank you that makes so much more sense - fixed